### PR TITLE
feat(training): upgrade base model to Qwen 3.5-4B

### LINKED ISSUE #140: Upgrade base model from Qwen2.5-3B to Qwen3.5-4B
**Parent:** #137

## Problem
Qwen2.5-3B struggles with JSON compliance and entity extraction at its parameter count. Qwen3.5-4B offers significant improvements: 256K context (vs 32K), better reasoning benchmarks, Gated Delta Networks for faster inference.

## Tasks
- [ ] Pull Qwen3.5-4B: `ollama pull qwen3.5:4b`
- [ ] Run baseline eval against Qwen3.5-4B without fine-tuning to measure zero-shot improvement
- [ ] Update training scripts to use Qwen3.5-4B as base model
- [ ] Update Unsloth target model configuration
- [ ] Verify LoRA adapter compatibility with Qwen3.5 architecture
- [ ] Re-train all 3 adapters on Qwen3.5-4B base
- [ ] Run eval harness and compare v1.0 (Qwen2.5-3B) vs v2.0 (Qwen3.5-4B)
- [ ] Update `scripts/training/config.py` with new base model defaults
- [ ] Update Modelfile templates for Ollama registration

## Acceptance Criteria
- [ ] Qwen3.5-4B baseline (no fine-tuning) outperforms Qwen2.5-3B fine-tuned on at least latency
- [ ] Fine-tuned Qwen3.5-4B meets ship criteria for at least query_parser
- [ ] Inference latency comparable or better than Qwen2.5-3B (despite +1B params, architecture is more efficient)

### PR DESCRIPTION
**Pull request by William-Hill**
## Summary

Closes #140

- Migrate Unsloth API from `FastLanguageModel` to `FastModel` and rename `tokenizer` → `processor` throughout train.py
- Update default base model from `unsloth/Qwen2.5-3B-Instruct` to `unsloth/Qwen3.5-4B`
- Bump sequence lengths (2048→4096, 4096→8192) and batch sizes for A100 40GB VRAM, keeping LoRA ranks unchanged for ablation discipline
- Version GGUF filenames with `qwen35` tag (e.g. `d4bl-query-parser-qwen35-q4_k_m.gguf`) across train.py, register_models.py, and Modelfiles

## Test Plan

- [x] All 268 training tests pass (`pytest tests/test_training/ --noconftest`)
- [x] Zero stale references to `FastLanguageModel`, `Qwen2.5-3B`, or old GGUF filenames in scripts/training/, models/, tests/test_training/
- [ ] Run training pipeline on Colab A100 to verify no OOM and model converges
- [ ] Run eval harness on exported GGUFs to compare against v1.0 baselines (0% hallucination accuracy, 59% entity F1)

## Summary by CodeRabbit

* **Documentation**
  * Added an implementation plan and design spec for the Qwen 3.5–4B upgrade.

* **New Features**
  * Resume/incremental support for data generation workflows.

* **Improvements**
  * Upgraded default training base model to Qwen 3.5–4B.
  * Increased context windows and sequence lengths across adapters.
  * Renamed and versioned model artifacts and deployment configs to include the Qwen3.5 suffix.

* **Tests**
  * Updated model config tests to reflect new context sizes and artifact names.

### CODE REVIEW

File: scripts/training/generate_training_pairs.py [hunk 1]
```diff
@@ -90,6 +90,47 @@ def _print_cost_summary() -> None:
...
+
+def _count_existing_lines(path: Path | None) -> int:
+    """Count lines in an existing JSONL file for resume support.
+
+    Returns 0 if the file doesn't exist or path is None.
+    """
+    if path is None or not path.exists():
+        return 0
+    with path.open() as fh:
+        return sum(1 for line in fh if line.strip())
```
Comment (William-Hill) [inline scripts/training/generate_training_pairs.py]: Re: all comments on `generate_training_pairs.py` resume logic — those changes were merged in PR #148 (training data expansion) and are not part of this PR's scope. This PR only modifies `train.py`, `register_models.py`, Modelfiles, and `test_modelfiles.py` for the Qwen 3.5-4B upgrade. The resume cursor concerns are valid and should be tracked as a separate issue.
